proposals/NativeInt.md

```diff
@@ -0,0 +1,263 @@
+| `|` | `native` | `native` | `native` | `or` |
+| `<<` | `native` | `int` | `native` | `shl` |
+| `>>` | `native` | `int` | `native` | `shr` / `shr.un` |
+
+For some binary operators, the IL operators support additional operand types (see [ECMA-335](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf)).
+But the set of operand types supported by C# is limited for simplicity and for consistency with existing operators in the language.
+
+Lifted versions of the operators, where the arguments and return types are `nint?` and `nuint?`, are supported.
+
+Compound assignment operations `x op= y` where `x` or `y` are native ints follow the same rules as with other primitive types with pre-defined operators.
+Specifically the expression is bound as `x = (T)(x op y)` where `T` is the type of `x` and where `x` is only evaluated once.
+
+### dynamic
+
+The conversions and operators are synthesized by the compiler and are not part of the underlying `IntPtr` and `UIntPtr` types.
+As a result those conversions and operators _are not available_ from the runtime binder for `dynamic`. 
+
+```C#
+nint x = 2;
+nint y = x + x; // ok
+dynamic d = x;
+nint z = d + x; // RuntimeBinderException: '+' cannot be applied 'System.IntPtr' and 'System.IntPtr'
+```
+
+### Type members
+
+The only constructor for `nint` or `nuint` is the parameter-less constructor.
+
+The members of `System.IntPtr` and `System.UIntPtr` other than the constructors and operators are available from `nint` and `nuint`.
+For `nint` the following `System.IntPtr` members are available. The members available on `nuint` are similar.
+
+```C#
+public static readonly IntPtr Zero;
+public static int Size { get; }
+public static IntPtr Add(IntPtr pointer, int offset);
+public static IntPtr Subtract(IntPtr pointer, int offset);
+public override bool Equals(object obj);
+public override int GetHashCode();
+public int ToInt32();
+public long ToInt64();
+public void* ToPointer();
+public override string ToString();
+public string ToString(string format);
+```
+
+Some members above use `IntPtr` in the signature even when used from `nint`.
+
+```C#
+nint x = nint.Zero;      // nint x
+nint y = nint.Add(x, 1); // nint y
+var z = nint.Zero;       // IntPtr z
+var w = nint.Add(x, 1);  // IntPtr w
+```
+
+Interfaces implemented by `System.IntPtr` and `System.UIntPtr` are available from `nint` and `nuint`.
+
+```C#
+nint n = 42;
+IEquatable<nint> i = n; // ok, IntPtr implements IEquatable<IntPtr>
+```
+
+### Overriding, hiding, and implementing
+
+The `native int` aliases and the underlying types are considered equivalent for overriding, hiding and implementing.
+
+Overloads cannot differ by `native int` and underlying type alone.
+Overrides and implementations may differ by `native int` and underlying type alone.
+Methods hide other methods that differ by `native int` and underlying type alone.
+
+### Miscellaneous
+
+`native int` types cannot be used as enum underlying types.
```

Comment:
Why not? Given the constraints for `constants` (values must be less than `int.MaxValue` and greater than `int.MinValue`) and that the runtime explicitly supports `native int` enums, it seems like this would be beneficial for some interop code?